site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java

```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
+                        try
+                        {
+                            AuthzGroup siteGroup = authzGroupService.getAuthzGroup(siteReference);
+                            Set<String> usersHasRole = siteGroup.getUsersHasRole(role);
+                            if (usersHasRole != null)
+                            {
+                                for( String userId : usersHasRole )
+                                {
+                                    Member member = site.getMember(userId);
+                                    group.addMember(userId, role, member.isActive(), false);
+                                }
+                            }
+                        }
+                        catch (GroupNotDefinedException e)
+                        {
+                            M_log.debug(this + ".processAutoCreateGroup: no authzgroup found for " + siteReference);
+                        }
+                    }
+                }else{
+                    createRandomGroupsForRole( roleList, intToSplit );
+                }
+            }
+
+            // save the changes
+            try
+            {
+                siteService.save(site);
+                // reset the form params
+                resetParams();
+            } 
+            catch (IdUnusedException e) {
+                M_log.warn(this + ".processAutoCreateGroup: cannot find site " + site.getId(), e);
+                return null;
+            } 
+            catch (PermissionException e) {
+                M_log.warn(this + ".processAutoCreateGroup: cannot find site " + site.getId(), e);
+                return null;
+            }
+        }
+
+        return "done";
+    }
+    
+    /**
+     * SAK-29373 - Create random groups of users from the given roster.
+     * This could be random number of users in specified number of groups,
+     * or vice versa.
+     * 
+     * @param rosters a list of provider IDs (should only ever be one in the list)
+     * @param unit the number of groups or users per group requested
+     */
+    private void createRandomGroupsForRoster( List<String> providerIDs, int unit )
+    {
+        String groupTitle = rosterGroupSplit ? rosterGroupTitleGroup: rosterGroupTitleUser;
+        if( !StringUtils.isEmpty( groupTitle ) )
+        {
+            List<String> userList = new ArrayList<>();
+            for( String providerID : providerIDs )
+            {
+                try
+                {
+                    Set<Membership> sectionMembers = cms.getSectionMemberships( providerID );
+                    for( Membership member : sectionMembers )
+                    {
+                        String userEID = member.getUserId();
+                        try
+                        {
+                            String userID = userDirectoryService.getUserId( userEID );
+                            if( !userList.contains( userID ) )
```

Comment:
this could get expensive with larger sections, consider profiling with some alternative implementations like keeping the userList sorted and doing a binary instead of linear search, may or may not be an overall win